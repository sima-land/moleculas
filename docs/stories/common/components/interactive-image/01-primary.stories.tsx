import { InteractiveImage, Parts } from '@sima-land/moleculas/common/components/interactive-image';
import { CSSProperties } from 'react';
import imageSrc from './image.png';

export const meta = {
  category: 'common/InteractiveImage',
  title: 'Простой пример',
  parameters: {
    layout: 'padded',
  },
};

interface TitledPoint {
  x: number;
  y: number;
  title: string;
}

const points: TitledPoint[] = [
  { x: 28, y: 30, title: 'Яйца' },
  { x: 63, y: 35, title: 'Кружка' },
  { x: 49, y: 70, title: 'Тарелка' },
  { x: 83, y: 69, title: 'Приборы' },
];

export default function Primary() {
  const style: CSSProperties = {
    borderRadius: '8px',
    width: '600px',
    maxWidth: '100%',
  };

  const onPointClick = (point: TitledPoint) => {
    alert(`Клик по точке "${point.title}"`);
  };

  return (
    <>
      <InteractiveImage style={style}>
        <Parts.Image src={imageSrc} />

        {points.map((point, index) => (
          <Parts.Point key={index} role='button' {...point} onClick={() => onPointClick(point)} />
        ))}
      </InteractiveImage>
    </>
  );
}
