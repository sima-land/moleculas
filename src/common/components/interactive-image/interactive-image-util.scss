@use 'pkg:@sima-land/ui-nucleons/breakpoints';

@mixin dot-size-default {
  --dot-size: 32px;
  --dot-size-core: 12px;
  @include breakpoints.down('xs') {
    --dot-size: 40px;
    --dot-size-core: 16px;
  }
}
