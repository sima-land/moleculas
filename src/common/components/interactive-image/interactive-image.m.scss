@use 'pkg:@sima-land/ui-nucleons/colors.scss';
@use 'pkg:@sima-land/ui-nucleons/breakpoints.scss';
@use './interactive-image-util';

.root {
  display: inline-block;
  font-size: 0;
  position: relative;
  overflow: hidden;
  background: colors.$basic-gray4;
  &:not(.dot-size-unset) {
    @include interactive-image-util.dot-size-default;
  }
}

.image-anchor,
.image {
  display: block;
  max-width: 100%;
  max-height: 100%;
  min-width: 100%;
  min-height: 100%;
}

.point {
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  width: var(--dot-size, 32px);
  height: var(--dot-size, 32px);
  background: rgba(colors.$basic-gray87, 0.4);
  border-radius: 50%;
  transform: translate(-50%, -50%);
  &:hover {
    cursor: pointer;
    background: rgba(colors.$basic-gray87, 0.64);
  }
  &::after {
    content: '';
    display: block;
    width: var(--dot-size-core, 12px);
    height: var(--dot-size-core, 12px);
    background: #fff;
    border-radius: 50%;
  }
}
