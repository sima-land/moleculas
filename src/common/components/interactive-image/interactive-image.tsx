import { AnchorHTMLAttributes, Children, forwardRef, isValidElement, ReactNode } from 'react';
import {
  InteractiveImageProps,
  InteractiveImageImageProps,
  InteractiveImagePointProps,
} from './types';
import classNames from 'classnames/bind';
import styles from './interactive-image.m.scss';

const cx = classNames.bind(styles);

export const InteractiveImage = forwardRef<HTMLDivElement, InteractiveImageProps>(
  ({ children, 'data-testid': testId, className, dotSize, ...rest }, ref) => {
    const rootClassName = cx(
      'root',
      {
        'dot-size-unset': dotSize === 'unset',
      },
      className,
    );

    validateChildren(children);

    return (
      <div ref={ref} className={rootClassName} {...rest} data-testid={testId}>
        {Children.toArray(children).filter(child => isValidElement(child))}
      </div>
    );
  },
);

/**
 * Проверка валидности содержимого, переданного в InteractiveImage.
 * @param children Содержимое.
 */
function validateChildren(children: ReactNode) {
  let counter = 0;

  for (const child of Children.toArray(children)) {
    if (isValidElement(child) && (child.type === ImageAnchor || child.type === Image)) {
      counter++;
    }
    if (counter > 1) {
      throw new Error('[InteractiveImage] Either Image or ImageAnchor allowed. Not both.');
    }
  }
}

/**
 * Обертка-ссылка для картинки.
 * @param props Свойства.
 * @return Элемент.
 */
function ImageAnchor({ children, className, ...rest }: AnchorHTMLAttributes<HTMLAnchorElement>) {
  return (
    <a
      {...rest}
      className={cx('image-anchor', className)}
      data-testid='interactive-image:image-anchor'
    >
      {Children.toArray(children).filter(child => isValidElement(child) && child.type === Image)}
    </a>
  );
}

/**
 * Изображение на котором будут располагаться точки.
 * @param props Свойства.
 * @return Элемент.
 */
const Image = forwardRef<HTMLImageElement, InteractiveImageImageProps>(
  ({ className, 'data-testid': testId = 'interactive-image:image', ...rest }, ref) => (
    <img ref={ref} className={cx('image', className)} data-testid={testId} {...rest} />
  ),
);

/**
 * Точка которая будет располагаться на изображении.
 * @param props Свойства.
 * @return Элемент.
 */
const Point = forwardRef<HTMLAnchorElement, InteractiveImagePointProps>(
  ({ x, y, className, style, 'data-testid': testId = 'interactive-image:point', ...rest }, ref) => (
    <a
      ref={ref}
      aria-label='Точка на изображении'
      data-testid={testId}
      className={cx('point', className)}
      style={{ ...style, top: `${y}%`, left: `${x}%` }}
      {...rest}
    />
  ),
);

export const Parts = { ImageAnchor, Image, Point } as const;
