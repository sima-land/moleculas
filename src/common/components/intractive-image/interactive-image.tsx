import React, {
  AnchorHTMLAttributes,
  Children,
  forwardRef,
  isValidElement,
  ReactNode,
} from 'react';
import {
  InteractiveImageProps,
  InteractiveImageImageProps,
  InteractiveImagePointProps,
} from './types';
import classNames from 'classnames/bind';
import styles from './interactive-image.module.scss';

const cx = classNames.bind(styles);

export const InteractiveImage = forwardRef<HTMLDivElement, InteractiveImageProps>(
  ({ children, 'data-testid': testId, className, dotSize, ...rest }, ref) => {
    const rootClassName = cx(
      'root',
      {
        'dot-size-unset': dotSize === 'unset',
      },
      className,
    );

    validateChildren(children);

    return (
      <div ref={ref} className={rootClassName} {...rest} data-testid={testId}>
        {Children.toArray(children).filter(
          child =>
            isValidElement(child) &&
            (child.type === ImageAnchor || child.type === Image || child.type === Point),
        )}
      </div>
    );
  },
);

function validateChildren(children: ReactNode) {
  let counter = 0;

  for (const child of Children.toArray(children)) {
    if (isValidElement(child) && (child.type === ImageAnchor || child.type === Image)) {
      counter++;
    }
    if (counter > 1) {
      throw new Error('[InteractiveImage] Either Image or ImageAnchor allowed. Not both.');
    }
  }
}

function ImageAnchor({ children, className, ...rest }: AnchorHTMLAttributes<HTMLAnchorElement>) {
  return (
    <a
      {...rest}
      className={cx('image-anchor', className)}
      data-testid='interactive-image:image-anchor'
    >
      {Children.toArray(children).filter(child => isValidElement(child) && child.type === Image)}
    </a>
  );
}

const Image = forwardRef<HTMLImageElement, InteractiveImageImageProps>(
  ({ className, 'data-testid': testId = 'interactive-image:image', ...rest }, ref) => (
    <img ref={ref} className={cx('image', className)} data-testid={testId} {...rest} />
  ),
);

const Point = forwardRef<HTMLAnchorElement, InteractiveImagePointProps>(
  ({ x, y, className, style, 'data-testid': testId = 'interactive-image:point', ...rest }, ref) => (
    <a
      ref={ref}
      aria-label='Точка на изображении'
      data-testid={testId}
      className={cx('point', className)}
      style={{ ...style, top: `${y}%`, left: `${x}%` }}
      {...rest}
    />
  ),
);

export const Parts = { ImageAnchor, Image, Point } as const;
