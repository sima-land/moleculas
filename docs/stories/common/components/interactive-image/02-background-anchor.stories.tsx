import { InteractiveImage, Parts } from '@sima-land/moleculas/common/components/interactive-image';
import { CSSProperties } from 'react';
import imageSrc from './image.png';

export default {
  title: 'common/InteractiveImage',
  component: InteractiveImage,
  parameters: {
    layout: 'padded',
  },
};

interface TitledPoint {
  x: number;
  y: number;
  title: string;
}

const points: TitledPoint[] = [
  { x: 28, y: 30, title: 'Яйца' },
  { x: 63, y: 35, title: 'Кружка' },
  { x: 49, y: 70, title: 'Тарелка' },
  { x: 83, y: 69, title: 'Приборы' },
];

export function BackgroundAnchor() {
  const style: CSSProperties = {
    borderRadius: '8px',
    width: '600px',
    maxWidth: '100%',
  };

  const onPointClick = (dot: TitledPoint) => {
    alert(`Клик по точке "${dot.title}"`);
  };

  return (
    <InteractiveImage style={style}>
      <Parts.ImageAnchor href='https://sima-land.ru' target='_blank'>
        <Parts.Image src={imageSrc} />
      </Parts.ImageAnchor>

      {points.map((dot, index) => (
        <Parts.Point key={index} role='button' {...dot} onClick={() => onPointClick(dot)} />
      ))}
    </InteractiveImage>
  );
}

BackgroundAnchor.storyName = 'Картинка-ссылка';
