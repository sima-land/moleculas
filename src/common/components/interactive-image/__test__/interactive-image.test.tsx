import { createRef } from 'react';
import { render } from '@testing-library/react';
import { InteractiveImage, Parts } from '../interactive-image';

describe('InteractiveImage', () => {
  it('Должен корректно обработать свойства и нарисовать компонент', () => {
    const { asFragment, queryAllByTestId } = render(
      <InteractiveImage>
        <Parts.Image src='https://www.images.com/123' />
        <Parts.Point role='button' x={1} y={2} />
        <Parts.Point role='button' x={2} y={3} />
        <Parts.Point role='button' x={3} y={4} />
        <Parts.Point role='button' x={4} y={5} />
      </InteractiveImage>,
    );

    expect(asFragment()).toMatchSnapshot();

    expect(queryAllByTestId('interactive-image:image')).toHaveLength(1);
    expect(queryAllByTestId('interactive-image:point')).toHaveLength(4);
  });

  it('Должен взять ссылку', () => {
    const ref = createRef<HTMLDivElement>();

    const { getByTestId } = render(
      <InteractiveImage ref={ref} data-testid='test-root'>
        <Parts.Image src='https://www.images.com/123' />
        <Parts.Point role='button' x={1} y={2} />
        <Parts.Point role='button' x={2} y={3} />
        <Parts.Point role='button' x={3} y={4} />
        <Parts.Point role='button' x={4} y={5} />
      </InteractiveImage>,
    );

    expect(ref.current).toBe(getByTestId('test-root'));
  });

  it('Должен корректно обработать свойство "dotSize"', () => {
    const { getByTestId } = render(
      <InteractiveImage dotSize='unset' data-testid='interactive-image'>
        <Parts.Image src='https://www.images.com/123' />
        <Parts.Point role='button' x={1} y={2} />
        <Parts.Point role='button' x={2} y={3} />
        <Parts.Point role='button' x={3} y={4} />
        <Parts.Point role='button' x={4} y={5} />
      </InteractiveImage>,
    );

    expect(getByTestId('interactive-image').classList.contains('dot-size-unset')).toBe(true);
  });

  it('Должен корректно нарисовать Parts.ImageAnchor', () => {
    const { queryAllByTestId, getByTestId } = render(
      <InteractiveImage>
        <Parts.ImageAnchor href='https://www.site.com' className='my-class'>
          <Parts.Image src='https://www.images.com/123' />
        </Parts.ImageAnchor>
        <Parts.Point role='button' x={1} y={2} />
        <Parts.Point role='button' x={2} y={3} />
        <Parts.Point role='button' x={3} y={4} />
      </InteractiveImage>,
    );

    expect(queryAllByTestId('interactive-image:image-anchor')).toHaveLength(1);
    expect(queryAllByTestId('interactive-image:image')).toHaveLength(1);
    expect(queryAllByTestId('interactive-image:point')).toHaveLength(3);

    const imageAnchor = getByTestId('interactive-image:image-anchor') as HTMLAnchorElement;
    expect(imageAnchor.href).toBe('https://www.site.com/');
    expect(imageAnchor.classList.contains('image-anchor')).toBe(true);
    expect(imageAnchor.classList.contains('my-class')).toBe(true);
  });

  it('Должна выдавать ошибку, когда и ImageAnchor, и Image находятся в дочерних элементах', () => {
    expect(() => {
      render(
        <InteractiveImage>
          <Parts.ImageAnchor href='https://www.site.com' className='my-class'>
            <Parts.Image src='https://www.images.com/123' />
          </Parts.ImageAnchor>
          <Parts.Image src='https://www.images.com/123' />
        </InteractiveImage>,
      );
    }).toThrow();
  });

  it('Должен корректно нарисовать точки в обертке', () => {
    const { queryAllByTestId } = render(
      <InteractiveImage>
        <Parts.Image src='https://www.images.com/123' />
        <div data-testid='wrapper'>
          <Parts.Point role='button' x={1} y={2} />
        </div>
      </InteractiveImage>,
    );

    expect(queryAllByTestId('interactive-image:image')).toHaveLength(1);
    expect(queryAllByTestId('wrapper')).toHaveLength(1);
    expect(queryAllByTestId('interactive-image:point')).toHaveLength(1);
  });
});
