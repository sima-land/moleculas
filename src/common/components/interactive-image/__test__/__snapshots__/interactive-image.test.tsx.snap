// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InteractiveImage Должен корректно обработать свойства и нарисовать компонент 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <img
      class="image"
      data-testid="interactive-image:image"
      src="https://www.images.com/123"
    />
    <a
      aria-label="Точка на изображении"
      class="point"
      data-testid="interactive-image:point"
      role="button"
      style="top: 2%; left: 1%;"
    />
    <a
      aria-label="Точка на изображении"
      class="point"
      data-testid="interactive-image:point"
      role="button"
      style="top: 3%; left: 2%;"
    />
    <a
      aria-label="Точка на изображении"
      class="point"
      data-testid="interactive-image:point"
      role="button"
      style="top: 4%; left: 3%;"
    />
    <a
      aria-label="Точка на изображении"
      class="point"
      data-testid="interactive-image:point"
      role="button"
      style="top: 5%; left: 4%;"
    />
  </div>
</DocumentFragment>
`;
